hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/HostCandidateIterator.java

```diff
@@ -0,0 +1,129 @@
+        this.hostManager = spawn.hostManager;
+        this.balancer = spawn.getSpawnBalancer();
+        this.sortedHosts = new TreeSet<>(hostAndScoreComparator);
+        this.taskScoreIncrement = balancer.getConfig().getSiblingWeight();
+
+        if(job != null) {
+            for (JobTask task : job.getCopyOfTasks()) {
+                for (String replicaHostId : task.getAllTaskHosts()) {
+                    HostState host = hostManager.getHostState(replicaHostId);
+                    Double score = scoreMap.getOrDefault(host, 0d);
+                    scoreMap.put(host, score + (double) this.taskScoreIncrement);
+                }
+            }
+        }
+        // create the sortedHosts priority queue which will be used for picking candidate hosts
+        for (Map.Entry<HostState, Double> entry : scoreMap.entrySet()) {
+            this.sortedHosts.add(new WithScore<>(entry.getKey(), entry.getValue()));
+        }
+
+        this.minScore = hostManager.getHostLocationSummary().getPriorityLevel().score;
+    }
+
+    /**
+     * Return a host chosen in order of zone preference, then score
+     */
+    public List<String> getNewReplicaHosts(int replicas, JobTask task,
+                                           @Nullable String excludeHostUuid, boolean isReplica) {
+        if(replicas <= 0) {
+            return new ArrayList<>();
+        }
+
+        Collection<HostLocation> locations = new ArrayList<>();
+        // get current host locations used by live/replicas
+        for (String replicaHostId : task.getAllTaskHosts()) {
+            if ((replicaHostId != null) && !replicaHostId.equals(excludeHostUuid)) {
+                locations.add(hostManager.getHostState(replicaHostId).getHostLocation());
+            }
+        }
+
+        List<String> chosenHostIds = new ArrayList<>(replicas);
+        // create each replica
+        for (int i = 0; i < replicas; i++) {
+            double minScoreSoFar = Double.MAX_VALUE;
+            WithScore<HostState> bestHost = null;
+            for (WithScore<HostState> hostAndScore : sortedHosts) {
+                HostState host = hostAndScore.element;
+                boolean excluded = host.getHostUuid().equals(excludeHostUuid);
+                if (excluded || !host.canMirrorTasks() || (isReplica && !balancer.okToPutReplicaOnHost(host, task))) {
+                    continue;
+                }
+                HostLocation hostLocation = host.getHostLocation();
+                int score = 0;
+                // HostLocation further away from task and replica locations gets lower score
+                for (HostLocation location : locations) {
+                    score += hostLocation.assignScoreByHostLocation(location);
+                }
+                double avgScore = locations.isEmpty()? 0 : (double) score / (double) locations.size();
+                if (avgScore == (double) this.minScore) {
```

Comment:
also don't understand how this works

Reply:
@kxingit I've added more documentation for the scores and replica selection strategy used here - is there something you're concerned about with replica selection?